### PR TITLE
`utils` and `utils.create_from_yaml` improvements

### PR DESCRIPTION
**Pull request by oz123**
As discussed in #722. 

Seems this had quite some positive feedback.

### CODE REVIEW

File: kubernetes/utils/create_from_yaml.py [hunk 1]
```diff
@@ -14,60 +14,119 @@
...
 from os import path
 
 import yaml
-from six import iteritems
 
 from kubernetes import client
 
 
-def create_from_yaml(k8s_client, yaml_file, verbose=False, **kwargs):
+def create_from_file(k8s_client, yaml_path, verbose=False, **kwargs):
```
Comment (juliantaylor) [inline kubernetes/utils/create_from_yaml.py]: create_from_stream using a file like object would be more flexible than create from file path
That way you support anything file like at once and not only files on disk.

File: kubernetes/utils/create_from_yaml.py [hunk 2]
```diff
@@ -14,60 +14,119 @@
...
+    Valid values are: - All: all dry run stages will be processed
+    """
+    with open(path.abspath(yaml_path)) as f:
+        k8s_api = create_from_yaml(k8s_client, f.read(), verbose=verbose,
+                                   **kwargs)
+
+    return k8s_api
+
+
+def create_from_yaml(k8s_client, yaml_str, verbose=False, **kwargs):
```
Comment (juliantaylor) [inline kubernetes/utils/create_from_yaml.py]: this function would be unnecessary if you add create_from stream as you could just add your yaml string to a io.StringIO() object

File: kubernetes/utils/create_from_yaml.py [hunk 3]
```diff
@@ -14,60 +14,119 @@
...
+
+    return k8s_api
+
+
+def create_from_map(k8s_client, yml_object, verbose, **kwargs):
     """
+    performe an action from a valid parsed yaml object (as dict).
+
+    :param k8s_client: an ApiClient objcet initialized with the client args
+    :yml_object dict: a parsed yaml object
```
Comment (juliantaylor) [inline kubernetes/utils/create_from_yaml.py]: I don't think one needs to explicitly reference yaml here as what the function wants is a kubernetes manifest in python dictionary form
though it probably also does not harm as yaml is the most common representation form of that data.

File: kubernetes/utils/create_from_yaml.py (hunk 2 above)
Comment (oz123) [reply kubernetes/utils/create_from_yaml.py]: That is true. I will send a fix later .

Comment (juliantaylor) [reply kubernetes/utils/create_from_yaml.py]: a concern is also backward compatibility
for that this function should probably be kept with the old behavior of using file paths

Though I am no developer I don't know the rules this project has for this.
it seems the 9.0 release has not yet been cut so maybe one can still change it before release

File: test-requirements.txt [hunk 4]
```diff
@@ -10,4 +10,6 @@ recommonmark
 codecov>=1.4.0
 pycodestyle
 autopep8
-isort
\ No newline at end of file
+isort
+jwt
+adal
```
Comment (roycaihw) [inline test-requirements.txt]: why do we change test requirements?

Comment (micw523) [reply test-requirements.txt]: I don't think we're using jwt or adal?

**Comment by micw523** [comment]
I think I'm still holding my original position in #722: we don't need many function names. We should be able to do everything in ```create_from_yaml``` (potentially with an extra input parameter), since the word ```yaml``` isn't specific to a file anyway. If absolutely needed we can make an alias on ```create_from_yaml```.

Also, we're pushing to merge 673, so it's likely that you'll need a rebase soon...

File: test-requirements.txt (hunk 4 above)
Comment (oz123) [reply test-requirements.txt]: @micw523, the test failed with import error without jwt and adal.

Comment (oz123) [reply test-requirements.txt]: I checked again:
```
$ grep -nr adal kubernetes | grep -v \.pyc
kubernetes/base/config/kube_config.py:24:import adal
kubernetes/base/config/kube_config.py:223:        context = adal.AuthenticationContext(
```

jwt is a dependency of adal, and I'm not sure adal is specifically required here.
I removed jwt. How do you want me to proceed with adal?

File: kubernetes/utils/create_from_yaml.py (hunk 2 above)
Comment (oz123) [reply kubernetes/utils/create_from_yaml.py]: @juliantaylor the function create_from_yaml now accepts a file or a string, which will be then parsed.

File: kubernetes/utils/create_from_yaml.py (hunk 1 above)
Comment (oz123) [reply kubernetes/utils/create_from_yaml.py]: I totally agree, but it seems that currently there will be no acceptance for that.

File: kubernetes/utils/create_from_yaml.py (hunk 3 above)
Comment (oz123) [reply kubernetes/utils/create_from_yaml.py]: this is why I added `create_from_map`

**Comment by micw523** [comment]
Which branch did you fork it from? If you forked it from the current master I’ll need to do some investigation.

Adal as a requirement got moved to optional in a previous PR so that might be the problem.

**Comment by oz123** [comment]
> Which branch did you fork it from? If you forked it from the current master I’ll need to do some investigation.
> 
> Adal as a requirement got moved to optional in a previous PR so that might be the problem.

I forked from the master on a fresh clone I did yesterday.

**Comment by micw523** [comment]
I just had a fresh fork from the master branch. adal is not required for testing. Is it breaking anything that you submitted?

**Comment by micw523** [comment]
I checked your past Travis builds. It doesn't seem like adal is affecting you. Please remove the added packages.

**Comment by oz123** [comment]
> I checked your past Travis builds. It doesn't seem like adal is affecting you. Please remove the added packages.

done. I completely removed that commit.

**Comment by oz123** [comment]
@micw523 the new master branch includes `create_from_yaml_single_item` which obsoletes my work.
I think a lot of people would appreciate a more explicit function name, because `create_from_yaml` is still misleading. A better name would have been `create_from_yaml_file`. In order to keep backward compatabilty `create_from_yaml_file` could be added in the future as a thin wrapper around `create_from_yaml_file` and if people use `create_from_yaml` use a warning, but still read the file and parse the yaml.
My two cents.

**Comment by micw523** [comment]
I’m not sure. The ```create_from_single_item``` is there for a multi-resource yaml and / or yaml lists. It is NOT imported into the package nor you can use it directly with inputs from a string if your yaml string / dict contains a list type or multiple resources. This idea is not obsolete if these cases are handled.

Do you still want to work on this, or you’re going to leave it in my hands?

**Comment by oz123** [comment]
@micw523 I would prefer starting on a clean branch and a clean PR. This has become too opaque for me. Do you mind if I submit a new PR?

**Comment by micw523** [comment]
That’s cool, go for it :)

**Pull request closed without merging**